playbooks/roles/compatibility-layer/tasks/main.yml

```diff
@@ -0,0 +1,28 @@
+# TODO: when startprefix does not exist
+# use container to install prefix?
+#- include_tasks: install_prefix.yml
+
+- name: Check if a Prefix installation is found at the specified location
+  stat:
+    path: "{{ prefix_location }}/usr/bin/emerge"
+  register: emerge
+
+- name: Start transaction
+  command: "cvmfs_server transaction {{ cvmfs_repository }}"
+  when: cvmfs_start_transaction
+
+- block:
+  - include_tasks: add_overlay.yml
+    when: emerge.stat.exists
```

Comment:
I agree that it would be really nice to repair a broken installation, but I'm not really sure how to do that, since it's hard to find out what broke it in the first place. 

The when clause now only checks if emerge is in your $PATH for the given Prefix path, so basically it only checks if this is a valid Prefix path. If that's not the case, it does not really make sense to continue anyway, so it just skips those included tasks. Maybe it should even abort the playbook in this case? That would allow us to remove the when clauses.

We can go a step further, for instance by doing the check that I described above in a reply to your comment about the `eselect repository add` command, i.e. by checking if our repository is part of `eselect-repo.conf` (and if it's correct). But what if there is another repo in there (which could have been added manually), which could in theory could cause issues? Should we just remove all other repos? I figured that, since this is a very EESSI-specific role anyway and we plan to do everything through this playbook (and only very occasionally), and we only have one custom overlay for now, this check was sufficient for now.

So I'm not really sure how to improve this / automatically fix things; do you have more concrete ideas about this?

Reply:
What I meant by broken is related to the imposed configuration or conditions on the ansible-playbook itself. So, there won't be some many things to fix, because the playbook doesn't know much about the gentoo-prefix internals. But that's completely ok.
I was more referring to tasks like the following one
```
- name: Make sets directory
   file:
     path: "{{ gentoo_prefix_path }}/etc/portage/sets"
     state: directory
     mode: '0755'
```
This task would be completely skipped if `emerge` was found in the PATH, but could potentially be problematic if someone by hand modifies the folder permissions. 

There is also another aspect. If someone updates the list of packages to install on the following task, after the prefix was installed, then there would be no point of running the Ansible playbook because the task to install the packages would be completely skipped. While the `portage` module is able to detect if a package was installed or not and would mark it as `ok` if it was already there.
```
  - include_tasks: install_packages.yml
     when: emerge.stat.exists
```

That's more or less what I was talking about the idempotency. So, we could run the ansible-playbook regularly to install new packages and guarantee that it is "sane" (within the ansible-playbook capabilities) at all times.

Reply:
Hmm, I don't really understand this, but perhaps we're saying the same thing now ;-)
The playbook will always run these tasks again (and hence, fix the permissions on that sets directory), unless it cannot find emerge at the given Prefix path. This is just a check to verify that the given path is an actual Prefix installation (which always should have emerge?). Without emerge, there's no point in running these tasks, I believe, since almost all of them require emerge or somehow depend on emerge. 
This is what you mean in the end of your comment, right?

Reply:
Not really. I think that if `emerge` is not found, then the tasks should fail and not be skipped.
Because that highlights a problem.
At the same time, changing the folder permissions for the `"{{ gentoo_prefix_path }}/etc/portage/sets"` folder is independent of the `emerge` command.

Reply:
Yes, failing makes sense (I think I proposed that before as well). I'll modify this and do the emerge check and fail it cannot be found. 

In that case it will still not run a few of the tasks if emerge cannot be found, but I don't think that makes sense anyway. If emerge is not available, I don't really see the point of, for instance, making/modifying that sets directory. This particular task is part of the process of adding an overlay and installing packages from that overlay, and I don't see the point in making that directory if you already know you cannot add the overlay and install packages from it.

Reply:
Done:

```
- name: Check if a Prefix installation is found at the specified location
  stat:
    path: "{{ gentoo_prefix_path }}/usr/bin/emerge"
  register: emerge

- name: Fail if the given Prefix path does not have emerge
  fail:
    msg: |
         The given Gentoo Prefix path does not seem to be valid,
         because the emerge command cannot be found in {{ gentoo_prefix_path }}/usr/bin.
         Please check it and fix (the path to) the Gentoo Prefix installation.
  when: not emerge.stat.exists
```

For an invalid path/installation, you now get this:
```
TASK [compatibility-layer : Fail if the given Prefix path does not have emerge] ******************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "The given Gentoo Prefix path does not seem to be valid,\nbecause the emerge command cannot be found in /tmp/usr/bin.\nPlease check it and fix (the path to) the Gentoo Prefix installation.\n"}
```

Reply:
> This particular task is part of the process of adding an overlay and installing packages from that overlay, and I don't see the point in making that directory if you already know you cannot add the overlay and install packages from it.

Ah! I reviewed it with a different mindset! Sorry about that.
In this case, yeap! You are right. I agree with you too. It is pointless to proceed with other tasks if the `emerge` command is not there.

Reply:
Okay, no problem! Thanks for the feedback.